Add CoffeeMachine implementation

EkaterinaFarygina opened the pull request:
**Task description**

Let's redesign our program and write a class that represents a coffee machine. The class should have a method that takes a string as input. Every time the user inputs a string to the console, the program invokes this method with one argument: the line that the user inputs to the console. This system simulates pretty accurately how real-world electronic devices work. External components (like buttons on the coffee machine or tapping on the screen) generate events that pass into the single interface of the program.

The class should not use system input at all; it will only handle the input that comes to it via this method and its string argument.

The first problem that comes to mind: how to write that method in a way that it represents all that coffee machine can do? If the user inputs a single number, how can the method determine what that number is: a variant of coffee chosen by the user or the number of the disposable cups that a special worker added into the coffee machine?

The right solution to this problem is to store the current state of the machine. The coffee machine has several states it can be in. For example, the state could be "choosing an action" or "choosing a type of coffee". Every time the user inputs something and a program passes that line to the method, the program determines how to interpret this line using the information about the current state. After processing this line, the state of the coffee machine can be changed or can stay the same.

Commit:
Add CoffeeMachine implementation

pavel-krizskii commented:
1. Let's rename `State` to `CoffeeMachineState` and change the values to `CHOOSING_AN_ACTION` and `CHOOSING_A_TYPE_OF_COFFEE` as suggested in the task.
2. Then let's add an instance field named `currentState` of type `CoffeeMachineState` to the `CoffeeMachine` class.
This instance field should be used when we process user's input. Depending on the input, it can be changed from `CHOOSING_AN_ACTION` to `CHOOSING_A_TYPE_OF_COFFEE` and vice versa.
3. However, we still need enum with possible actions (`BUY`, `FILL`, `REMAINING`, `TAKE`). Let's create another enum called `CoffeeMachineAction` that will hold these constants.
4. Since `Scanner` should not be used inside `CoffeeMachine`, let's read user's input from another class. We can call it `CoffeeMachineCommandsScanner`. All commands should be read in this class. To understand which message to display to the user: `Write action (buy, fill, take, remaining, exit):` or `What do you want to buy?...` we can check current `CoffeeMachineState`, so the `CoffeeMachineCommandsScanner` should have an access to the `CoffeeMachine` instance.

Commit:
Correct CoffeeMachine

1. State renamed to CoffeeMachineState with new states
2. Added CoffeeMachineAction with possible actions for processing
3. Added CoffeeMachineCommandScanner, which asks for user input
3. Added new instance field to CoffeeMachine - CoffeeMachineState
4. Renamed method coffeeMachine to processUserInput + Changed logic
5. Corrected access modifiers to more restrictive

pavel-krizskii commented on jetbrains_academy_introduction_to_java.iml:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
`.iml` files in general should not be committed. These are internal files used by IntelliJ Idea.
Better to exclude them from versioning by adding a corresponding rule to `.gitignore`

pavel-krizskii commented on src/machine/Coffee.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    int getAmountOfWater() {
+        return this.amountOfWater;
```
Usually `this` keyword is not used in getters. It is usually used when there is a method/constructor parameter or a local variable with the same name as an instance field

pavel-krizskii commented on src/machine/Coffee.java:
```diff
@@ -0,0 +1,41 @@
+    private final int amountOfMilk;
+    private final int amountOfCoffeeBeans;
+    private final int amountOfDisposableCups;
+    private final int price;
```
It is not clear which currency is used here for the `price` and what to do if the price is not an integer.

Ideally, we should use some `Money` class instead of `int`. This `Money` class could have `Currency` instance field and some other instance field(s) representing the price itself.

However, implementing a good `Money` class is not in the scope of this task.
But at least, we could rename the field to `priceInDollars` and leave a small comment that the price will always be an integer.

pavel-krizskii commented on src/machine/CoffeeMachine.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    CoffeeMachine(int amountOfWater, int amountOfMilk, int amountOfCoffeeBeans,
+                         int amountOfDisposableCups, int money, CoffeeMachineState currentState) {
```
1. It is better to set initial `currentState` right inside the constructor. Ideally this state should be changed/set only from inside the `CoffeeMachine`
2. minor: fix the alignment

pavel-krizskii commented on src/machine/CoffeeMachine.java:
```diff
@@ -0,0 +1,111 @@
+    private int amountOfCoffeeBeans;
+    private int amountOfDisposableCups;
+    private int money;
+    private CoffeeMachineState currentState;
```
minor: we can add an empty line before the `currentState ` to visually separate it from the other fields that represent the amount of remaining supplies.

Also we can think about extracting the supplies to a separate class, but it's not necessary at this stage

pavel-krizskii commented on src/machine/CoffeeMachine.java:
```diff
@@ -0,0 +1,111 @@
+    private void prepareCoffee(int chosenOption) {
+        Coffee coffee = null;
+        switch (chosenOption) {
+            case 1 -> coffee = Coffee.ESPRESSO;
```
We can add static method to the `Coffee` enum: `Coffee fromOption(int option)` and move this switch there.

Then here we can write: `final var coffee = Coffee.fromOption(chosenOption)`

pavel-krizskii commented on src/machine/CoffeeMachine.java:
```diff
@@ -0,0 +1,111 @@
+        }
+        if (coffee != null) {
+            if (!isEnoughIngredients(coffee)) return;
+            this.amountOfWater -= coffee.getAmountOfWater();
+            this.amountOfMilk -= coffee.getAmountOfMilk();
+            this.amountOfCoffeeBeans -= coffee.getAmountOfCoffeeBeans();
+            this.amountOfDisposableCups -= coffee.getAmountOfDisposableCups();
+            this.money += coffee.getPrice();
```
minor: better to extract it to a separate private method

pavel-krizskii commented on src/machine/CoffeeMachine.java:
```diff
@@ -0,0 +1,111 @@
+        }
+    }
+
+    private void fillSupplies() {
```
Looks like we could add another coffee machine state `FILLING_SUPPLIES`.

Then we should get rid of this `Scanner` inside the `CoffeeMachine` and read everything inside `CoffeeMachineCommandsScanner`

pavel-krizskii commented on src/machine/CoffeeMachine.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    void processUserInput(String input) {
+        if (input.equals("1") || input.equals("2") || input.equals("3")) {
```
minor: usually inside the `equals` method there is a check for `null`. That's why when we want to compare some variable/method argument with some constant, we write `"constant".equals(variable)`. This way, if the `variable` is null, we won't get an NPE and simply get `false`

pavel-krizskii replied:
Let's use a switch-case here on `currentState`

pavel-krizskii commented on src/machine/CoffeeMachineCommandsScanner.java:
```diff
@@ -0,0 +1,35 @@
+        CoffeeMachine coffeeMachine = new CoffeeMachine(400, 540, 120,
+                9, 550, coffeeMachineState);
+        while (coffeeMachine.getCurrentState() != CoffeeMachineState.OFF) {
+            if (coffeeMachine.getCurrentState() == CoffeeMachineState.CHOOSING_AN_ACTION) {
```
minor: it is a bit better to use switch-case instead of if-else when we check the enum values

pavel-krizskii commented on src/machine/CoffeeMachineCommandsScanner.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    private static void askForTypeOfCoffee(CoffeeMachine coffeeMachine) {
+        Scanner scanner = new Scanner(System.in);
```
We could reuse the same `Scanner` instance instead of creating a new one every time

pavel-krizskii commented on src/machine/CoffeeMachineState.java:
```diff
@@ -0,0 +1,6 @@
+
+enum CoffeeMachineState {
+    CHOOSING_AN_ACTION, CHOOSING_A_TYPE_OF_COFFEE, OFF
+
```
minor: remove empty lines

Commit:
Correct CoffeeMachine

1. Corrections according to comments
2. Add Supplies class, which stores supplies, available in CoffeeMachine

pavel-krizskii commented on src/machine/Coffee.java:
```diff
@@ -0,0 +1,51 @@
+    private final int amountOfMilk;
+    private final int amountOfCoffeeBeans;
+    private final int amountOfDisposableCups;
+    // It's ok, that the type of the price is int for this task
```
Maybe `For our purposes it is sufficient that the price has int type`?

pavel-krizskii commented on src/machine/CoffeeMachine.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    CoffeeMachine() {
+        supplies = new Supplies(400, 540, 120, 9);
```
It would be better to have an opportunity to pass the initial supplies and money to the constructor

pavel-krizskii commented on src/machine/CoffeeMachine.java:
```diff
@@ -0,0 +1,92 @@
+
+    private void displayContent() {
+        System.out.println("\nThe coffee machine has:");
+        System.out.printf("%d ml of water\n", supplies.getAmountOfWater());
+        System.out.printf("%d ml of milk\n", supplies.getAmountOfMilk());
+        System.out.printf("%d g of coffee beans\n", supplies.getAmountOfCoffeeBeans());
+        System.out.printf("%d disposable cups\n", supplies.getAmountOfDisposableCups());
```
Let's move these prints to `displayContent()` of `Supplies`

pavel-krizskii commented on src/machine/CoffeeMachine.java:
```diff
@@ -0,0 +1,92 @@
+
+    private void prepareCoffee(int chosenOption) {
+        final var coffee = Coffee.fromOption(chosenOption);
+        if (coffee != null) {
```
A bit shorter way: `if (coffee == null || !isEnoughIngredients(coffee)) return;`

Also this way we don't need to add extra 4 spaces for the following lines

pavel-krizskii commented on src/machine/CoffeeMachine.java:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    private boolean isEnoughIngredients(Coffee coffee) {
+        if (supplies.getAmountOfWater() < coffee.getAmountOfWater()) {
+            System.out.println("Sorry, not enough water!");
+            return false;
+        } else if (supplies.getAmountOfMilk() < coffee.getAmountOfMilk()) {
+            System.out.println("Sorry, not enough milk!");
+            return false;
+        } else if (supplies.getAmountOfCoffeeBeans() < coffee.getAmountOfCoffeeBeans()) {
+            System.out.println("Sorry, not enough coffee beans!");
+            return false;
+        } else if (supplies.getAmountOfDisposableCups() < coffee.getAmountOfDisposableCups()) {
+            System.out.println("Sorry, not enough disposable cups!");
+            return false;
```
Let's add a new method to `Supplies`: `boolean canMake(Coffee coffee)` and move these checks there

pavel-krizskii commented on src/machine/CoffeeMachine.java:
```diff
@@ -0,0 +1,92 @@
+                System.out.println();
+                this.currentState = CoffeeMachineState.CHOOSING_AN_ACTION;
+            }
+            default -> {
```
Let's add another case here for `CoffeeMachineState.CHOOSING_AN_ACTION` to make it more explicit

Commit:
Correct CoffeeMachine

1. Corrections according to comments

pavel-krizskii commented on src/machine/CoffeeMachine.java:
```diff
@@ -0,0 +1,80 @@
+    private final Supplies supplies;
+
+    // For our purposes it is sufficient that the price has int type
+    private int money;
```
Let's rename it to `dollars`? 

Also would be nice if the `For our purposes it is sufficient that the money has int type` comment was the same here and in the `Coffee` enum

pavel-krizskii commented on src/machine/CoffeeMachine.java:
```diff
@@ -0,0 +1,80 @@
+        money += coffee.getPriceInDollars();
+    }
+
+    private void takeMoney() {
```
minor: let's highlight in the name that we take all money from the coffee machine

pavel-krizskii replied:
Would `takeAllMoney` be a better name?

pavel-krizskii commented on src/machine/CoffeeMachine.java:
```diff
@@ -0,0 +1,80 @@
+                if ("back".equals(input)) {
+                    currentState = CoffeeMachineState.CHOOSING_AN_ACTION;
+                } else {
+                    int chosenOption = Integer.parseInt(input);
```
minor: `final int chosenOption = ...`

pavel-krizskii replied:
What if the input is not a valid integer? Let's handle this situation

pavel-krizskii commented on src/machine/CoffeeMachine.java:
```diff
@@ -0,0 +1,80 @@
+                } else {
+                    int chosenOption = Integer.parseInt(input);
+                    prepareCoffee(chosenOption);
+                    System.out.println();
```
It is not clear why we have `System.out.println();` here. Can we move it somewhere else closer to other printlns?

pavel-krizskii commented on src/machine/CoffeeMachine.java:
```diff
@@ -0,0 +1,80 @@
+                }
+            }
+            case CHOOSING_AN_ACTION -> {
+                CoffeeMachineAction coffeeMachineAction = CoffeeMachineAction.valueOf(input.toUpperCase());
```
What if the input is not a valid action? Let's handle this situation

pavel-krizskii replied:
minor: `final var coffeeMachineAction = ...` would be a bit better

pavel-krizskii commented on src/machine/CoffeeMachineCommandsScanner.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public static void main(String[] args) {
+        Supplies supplies = new Supplies(400, 540, 120, 9);
+        CoffeeMachine coffeeMachine = new CoffeeMachine(supplies, 550);
+        Scanner scanner = new Scanner(System.in);
```
We can create the scanner in the constructor of `CoffeeMachineCommandsScanner`, then we would not need to pass it to every method

pavel-krizskii replied:
Using `final var name` is a bit better for two reasons: we make it clear that the variable won't change and all variable names have the same alignment.

e.g.

```
final var supplies = ...
final var coffeeMachine = ...
final var scanner = ...
```

pavel-krizskii commented on src/machine/CoffeeMachineState.java:
```diff
@@ -0,0 +1,5 @@
+package machine;
+
+enum CoffeeMachineState {
+    CHOOSING_AN_ACTION, CHOOSING_A_TYPE_OF_COFFEE, FILLING_SUPPLIES, OFF
```
minor: better to use new lines here. Usually one line is used when it's clear that the enum cannot be extended in the future, but here looks like we might want to add more states

pavel-krizskii commented on src/machine/Supplies.java:
```diff
@@ -0,0 +1,70 @@
+package machine;
+
+public class Supplies {
```
minor: better to rename it to `CoffeeMachineSupplies` so all class names start with `Coffee` ☕

pavel-krizskii commented on src/machine/Supplies.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    void reduceSupplies(Coffee coffee) {
+        amountOfWater -= coffee.getAmountOfWater();
+        amountOfMilk -= coffee.getAmountOfMilk();
+        amountOfCoffeeBeans -= coffee.getAmountOfCoffeeBeans();
+        amountOfDisposableCups -= coffee.getAmountOfDisposableCups();
```
Let's validate that the supplies can be reduced here. We can throw an exception if they cannot. Hint: probably some custom exception class would be needed

pavel-krizskii commented on jetbrains_academy_introduction_to_java.iml:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<module type="JAVA_MODULE" version="4">
```
The file is still here 😿

pavel-krizskii replied:
Still not removed

pavel-krizskii commented on src/machine/CoffeeMachine.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    private boolean isEnoughIngredients(Coffee coffee) {
+        if (!supplies.canMakeCoffee(coffee)) {
```
minor: it can be named just `canMake` since it has `coffee` as the 1st parameter

pavel-krizskii commented on src/machine/CoffeeMachine.java:
```diff
@@ -0,0 +1,80 @@
+
+    private void prepareCoffee(int chosenOption) {
+        final var coffee = Coffee.fromOption(chosenOption);
+        if (coffee == null || !isEnoughIngredients(coffee)) return;
```
minor: `return` would be more noticeable on the new line

Commit:
Correct CoffeeMachine

1. Corrections according to comments

pavel-krizskii commented on src/machine/CoffeeMachineCommandsScanner.java:
```diff
@@ -0,0 +1,52 @@
+    public static void main(String[] args) {
+        final var coffeeMachineSupplies = new CoffeeMachineSupplies(400, 540, 120, 9);
+        final var coffeeMachine = new CoffeeMachine(coffeeMachineSupplies, 550);
+        final var coffeeMachineCommandsScanner = new CoffeeMachineCommandsScanner();
```
minor: we can name this local variable just `commandsScanner` to make the code shorter, since this variable is used quite often below

pavel-krizskii commented on src/machine/CoffeeMachineSupplies.java:
```diff
@@ -0,0 +1,82 @@
+        return amountOfDisposableCups;
+    }
+
+    private boolean isEnoughIngredients(Coffee coffee) {
+        if (!canMake(coffee)) {
+            return false;
+        } else {
+            System.out.println("I have enough resources, making you a coffee!");
+            return true;
+        }
+    }
```
This method should not be placed in this class, since `CoffeeMachineSupplies` doesn't know anything about coffee preparation :)

EkaterinaFarygina replied:
`isEnoughIngredients(Coffee coffee)` seems like redundant, because we have `canMake(Coffee coffee)` with main logic. Only `System.out.println("I have enough resources, making you a coffee!");` is missing.

pavel-krizskii commented on src/machine/CoffeeMachine.java:
```diff
@@ -0,0 +1,80 @@
+            coffeeMachineSupplies.reduceSupplies(coffee);
+            dollars += coffee.getPriceInDollars();
+            System.out.println();
+        } catch (IllegalStateException exception) {}
```
Should we log here that the coffee cannot be prepared?

EkaterinaFarygina replied:
We already have output for user for every case in `canMake(Coffee coffee)` inside `coffeeMachineSupplies.reduceSupplies(coffee)`

pavel-krizskii commented on src/machine/CoffeeMachine.java:
```diff
@@ -0,0 +1,80 @@
+    void processCommand(String input) {
+        switch (currentState) {
+            case CHOOSING_A_TYPE_OF_COFFEE -> {
+                if ("back".equals(input)) {
+                    currentState = CoffeeMachineState.CHOOSING_AN_ACTION;
+                } else {
+                    if (!input.matches("[1-3]")) {
+                        System.out.println("There is no such option. Please try again");
+                        return;
+                    }
+                    final int chosenOption = Integer.parseInt(input);
+                    prepareCoffee(chosenOption);
+                    this.currentState = CoffeeMachineState.CHOOSING_AN_ACTION;
+                }
```
Wdyt about extracting this case body and the below one to separate private methods?

Commit:
Correct CoffeeMachine

1. Corrections according to comments

Commit:
Correct CoffeeMachine

1. Corrections according to comments

Commit:
Change package name from machine to CoffeeMachine

pavel-krizskii reviewed (approved):
LGTM
